### PR TITLE
Improve UT coverage for IBMPowerVSCluster deletion flow

### LINKED ISSUE #1818: EPIC: Add UT accross various files in repository
/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Recently we added lots of functionality but haven't added any corresponding UT's to it. Its time to add unit test cases and overall goal is to reach 80% of test coverage.

As a non-goal during the time of adding UT if we find any redundant or unwanted or code duplication, a separate PR needs to be submitted to optimize the code.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

- [ ] cloud/scope/powervs_cluster.go 
- [ ] cloud/scope/powervs_machine.go 

Instructions for running individual tests on VS Code

1. Get kubebuilder asset path
```
$ cd cluster-api-provider-ibmcloud
$ make setup-envtest
kube-builder assets set using darwin OS at location /Users/karthikkn/Library/Application Support/io.kubebuilder.envtest/k8s/1.29.3-darwin-amd64
```

2. Create settings.json file with kubebuilder asset path
```
$ cd cluster-api-provider-ibmcloud/.vscode
$ cat settings.json
cat settings.json
{
    "go.testEnvVars": {
        "KUBEBUILDER_ASSETS": "/Users/karthikkn/Library/Application Support/io.kubebuilder.envtest/k8s/1.29.3-darwin-amd64",
    },
}

```
3. Run individual test by clicking on run test above test function name in VS Code editor

Instructions for checking code coverage of file 

1. Generate the test coverage html page
```
$ cd cluster-api-provider-ibmcloud/.vscode
$ make test-cover
```
A file named cover.html will be created under cluster-api-provider-ibmcloud directory

example path: /Users/karthikkn/k8-workspace/cluster-api-provider-ibmcloud/cover.html

2. Copy the path of the file and paste in browser to view the html page and select the file to view the lines covered in UT.

image

```[tasklist]
### Helper PRs
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1913
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1899
```

```[tasklist]
### UT PRs
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1912
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1927
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1928
```

### PR DESCRIPTION
**Pull request by Shilpa-Gokul**
This PR covers UT for deleteTransitGatewayConnections, DeleteDHCPServer, DeleteServiceInstance, DeleteCOSInstance, isResourceCreatedByController functions

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1818 

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: cloud/scope/powervs_cluster_test.go [hunk 1]
```diff
@@ -40,6 +41,7 @@ import (
 	mockRC "sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/resourcecontroller/mock"
 	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/resourcemanager"
 	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/transitgateway"
+	mockTG "sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/transitgateway/mock"
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:44]:
```suggestion
	tgmock "sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/transitgateway/mock"
```
to keep it consistent with other files

File: cloud/scope/powervs_cluster_test.go [hunk 2]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+		clusterScope   PowerVSClusterScope
+		expectedResult bool
+	}{
+		{
+			name: "When resourceType is VPC and VPC status is nil",
+			clusterScope: PowerVSClusterScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{},
+			},
+			resourceType:   infrav1beta2.ResourceTypeVPC,
+			expectedResult: false,
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3849]: bool will be default be false. so we can remove it or keep it for better readability

File: cloud/scope/powervs_cluster_test.go [hunk 3]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+					},
+				},
+			},
+			resourceType:   infrav1beta2.ResourceTypeVPC,
+			expectedResult: true,
+		},
+		{
+			name: "When resourceType is ServiceInstance and ServiceInstance status is nil",
+			clusterScope: PowerVSClusterScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{},
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:3868]: I believe this declaration is to avoid panic?

File: cloud/scope/powervs_cluster_test.go [hunk 4]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					COSInstance: &infrav1beta2.ResourceReference{
+						ID:                ptr.To("cosInstanceID"),
+						ControllerCreated: ptr.To(true),
+					},
+				},
+			},
+			ResourceClient: mockResourceController,
+		}
+		mockResourceController.EXPECT().GetResourceInstance(gomock.Any()).Return(nil, &core.DetailedResponse{StatusCode: ResourceNotFoundCode}, fmt.Errorf("GetResourceInstance error"))
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:4046]:
```suggestion
		mockResourceController.EXPECT().GetResourceInstance(gomock.Any()).Return(nil, &core.DetailedResponse{StatusCode: ResourceNotFoundCode}, fmt.Errorf("error getting resource instance"))
```

Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go:4046]: lets keep errors as uniform like other

File: cloud/scope/powervs_cluster_test.go (hunk 3 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:3868]: Yes

File: cloud/scope/powervs_cluster_test.go (hunk 2 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:3849]: I have kept it as is for easy readability

File: cloud/scope/powervs_cluster_test.go (hunk 1 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Addressed

**Comment by Amulyam24** [comment]
/retitle Improve UT coverage for IBMPowerVSCluster deletion flow

File: cloud/scope/powervs_cluster_test.go [hunk 5]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockResourceController = mockRC.NewMockResourceController(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When cos instance resource is not created by controller", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:3985]:
```suggestion
	t.Run("When COS instance resource is not created by controller", func(t *testing.T) {
```
Better to keep the names uniform throughout

File: cloud/scope/powervs_cluster_test.go [hunk 6]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When PowerVS service instance resource is not created by controller", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{}}
+		isDeleted, err := clusterScope.DeleteServiceInstance()
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:4136]:
```suggestion
		requeue, err := clusterScope.DeleteServiceInstance()
```
Please change wherever applicable

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm
/assign @Amulyam24

File: cloud/scope/powervs_cluster_test.go [hunk 7]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockTransitGateway = tgmock.NewMockTransitGateway(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When transit gateway in PowerVS is in deleting state", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:4408]:
```suggestion
	t.Run("When PowerVS connection of transit gateway is in deleting state", func(t *testing.T) {
```
The test case name might cause confusion, please change it wherever applicable for this function to specify as `PVS connection or VPC connection of a TG`

File: cloud/scope/powervs_cluster_test.go [hunk 8]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+			},
+			TransitGatewayClient: mockTransitGateway,
+		}
+		tg := &tgapiv1.TransitGateway{ID: ptr.To("transitGatewayID")}
+		mockTransitGateway.EXPECT().GetTransitGatewayConnection(gomock.Any()).Return(nil, &core.DetailedResponse{StatusCode: 400}, fmt.Errorf("error getting transit gateway connection"))
+		requeue, err := clusterScope.deleteTransitGatewayConnections(tg)
+		g.Expect(err.Error()).To(Equal("failed to get transit gateway powervs connection: error getting transit gateway connection"))
+		g.Expect(requeue).To(BeFalse())
+	})
+	t.Run("When transit gateway in PowerVS is not found", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:4509]:
```suggestion
	t.Run("When PowerVS connection is not found and VPC connection is deleted successfully", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 9]
```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
...
+		powerVSTGOptions := &tgapiv1.GetTransitGatewayConnectionOptions{TransitGatewayID: tg.ID, ID: ptr.To("powerVStgID")}
+		mockTransitGateway.EXPECT().GetTransitGatewayConnection(powerVSTGOptions).Return(nil, &core.DetailedResponse{StatusCode: ResourceNotFoundCode}, nil)
+		vpcTGOptions := &tgapiv1.GetTransitGatewayConnectionOptions{TransitGatewayID: tg.ID, ID: ptr.To("vpctgID")}
+		mockTransitGateway.EXPECT().GetTransitGatewayConnection(vpcTGOptions).Return(tgResponse, &core.DetailedResponse{StatusCode: 200}, nil)
+		mockTransitGateway.EXPECT().DeleteTransitGatewayConnection(gomock.Any()).Return(nil, nil)
+		requeue, err := clusterScope.deleteTransitGatewayConnections(tg)
+		g.Expect(err).To(BeNil())
+		g.Expect(requeue).To(BeTrue())
+	})
+	t.Run("When GetTransitGatewayConnection in VPC returns error", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:4542]:
```suggestion
	t.Run("When GetTransitGatewayConnection for VPC connection returns error", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go (hunk 5 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 6 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 8 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 9 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go]: Done

**Review by Amulyam24 — APPROVED** [review]
/lgtm

Thank you @Shilpa-Gokul!

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Pull request merged**